plan/statscache/statscache.go

```diff
@@ -14,89 +14,109 @@
 package statscache
 
 import (
+	"fmt"
 	"sync"
-	"time"
 
+	"github.com/juju/errors"
 	"github.com/ngaut/log"
 	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/meta"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/plan/statistics"
-	"github.com/pingcap/tidb/store/tikv/oracle"
+	"github.com/pingcap/tidb/util/sqlexec"
 )
 
 type statsInfo struct {
-	tbl      *statistics.Table
-	loadTime int64
+	m       sync.RWMutex
+	tbl     *statistics.Table
+	version uint64
 }
 
-type statsCache struct {
-	m     sync.RWMutex
-	cache map[int64]statsInfo
+// Handle can update stats info periodically.
+type Handle struct {
+	ctx         context.Context
+	lastVersion uint64
 }
 
-var statsTblCache = statsCache{cache: map[int64]statsInfo{}}
-
-// expireDuration is 1 hour.
-var expireDuration int64 = 60 * 60 * 1000
-
-func tableCacheExpired(si statsInfo) bool {
-	duration := oracle.GetPhysical(time.Now()) - si.loadTime
-	if duration >= expireDuration {
-		return true
-	}
-	return false
+// NewHandle creates a Handle for update stats.
+func NewHandle(ctx context.Context) *Handle {
+	return &Handle{ctx: ctx}
 }
 
-// GetStatisticsTableCache retrieves the statistics table from cache, and will refill cache if necessary.
-func GetStatisticsTableCache(ctx context.Context, tblInfo *model.TableInfo) *statistics.Table {
-	statsTblCache.m.RLock()
-	si, ok := statsTblCache.cache[tblInfo.ID]
-	statsTblCache.m.RUnlock()
-	// Here we check the TableInfo because there may be some ddl changes in the duration period.
-	// Also, we rely on the fact that TableInfo will not be same if and only if there are ddl changes.
-	if ok && tblInfo == si.tbl.Info && !tableCacheExpired(si) {
-		return si.tbl
+// Update reads stats meta from store and updates the stats map.
+func (h *Handle) Update(m *meta.Meta, is infoschema.InfoSchema) error {
+	sql := fmt.Sprintf("SELECT version, table_id from mysql.stats_meta where version > %d order by version", h.lastVersion)
+	rows, _, err := h.ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(h.ctx, sql)
+	if err != nil {
+		return errors.Trace(err)
 	}
-	txn := ctx.Txn()
-	if txn == nil {
-		err := ctx.ActivePendingTxn()
-		if err != nil || ctx.Txn() == nil {
-			return statistics.PseudoTable(tblInfo)
+	for _, row := range rows {
+		version, tableID := row.Data[0].GetUint64(), row.Data[1].GetInt64()
+		table, ok := is.TableByID(tableID)
+		if !ok {
+			log.Warnf("Unknown table ID %d in stats meta table, maybe it has been dropped", tableID)
+			continue
+		}
+		tableInfo := table.Meta()
+		tpb, err := m.GetTableStats(tableID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		var tbl *statistics.Table
+		if tpb != nil {
+			tbl, err = statistics.TableFromPB(tableInfo, tpb)
+			// Error is not nil may mean that there are some ddl changes on this table, so the origin
+			// statistics can not be used any more, we give it a pseudo one and save in cache.
```

Comment:
Here the pseudo one is not easily see, keep it nil is not the same as give it a pseudo one.